### PR TITLE
Port tests to vitest and add new tests

### LINKED ISSUE #2180: Add tests for MRC
**User Story:**
As a developer
I want to know the application works without manual testing
So that I can ship faster and be happy.

Test priority:
for stores, test in isolation, outside of component. Treat like a state machine. 
- [x] cartStore
- [x] Permit
- [x] CheckoutStore
- [ ] Rework tests for existing pages using new state management.
- [ ] Test new checkout page. DON'T mock the store. only mock wallet interactions, or what is absolutely necessary.
- [x] Test all view-only components.

### PR DESCRIPTION
**Pull request by vacekj**
fixes #2180

### CODE REVIEW

File: packages/grant-explorer/src/features/api/voting.ts [hunk 1]
```diff
@@ -13,6 +22,25 @@ export type PermitSignature = {
...
+export type PermitType = "dai" | "eip2612";
+/** Given a payout token, selects the correct permit type.
+ * - DAI is the old permit type without `value` and with the `allowed` prop
+ * - eip2612 is the standard permit interface, as specified in https://eips.ethereum.org/EIPS/eip-2612
+ *
+ * Old DAI permit type is only implemented on Ethereum and Polygon PoS. Check /docs/DAI.md for more info.
+ * */
+export const selectPermitType = (token: PayoutToken): PermitType => {
+  if (
+    /DAI/i.test(token.name) &&
```
Comment (bard) [inline packages/grant-explorer/src/features/api/voting.ts:34]: Do we already know all the tokens that need this? I'd prefer a list here, "any token that contains the string `dai`" seems a fairly lax check.

File: packages/grant-explorer/src/features/api/voting.ts [hunk 2]
```diff
@@ -232,3 +244,26 @@ export const signPermitDai = async ({
...
+export function encodeQFVotes(
+  donationToken: PayoutToken,
+  donations: Pick<
+    CartProject,
+    "amount" | "recipient" | "projectRegistryId" | "applicationIndex"
+  >[]
+): Hex[] {
+  return donations.map((donation) => {
+    const vote = [
+      getAddress(donationToken.address) as Hex,
```
Comment (bard) [inline packages/grant-explorer/src/features/api/voting.ts:257]: https://github.com/gitcoinco/grants-stack/issues/2158

;)

File: packages/grant-explorer/src/features/common/__tests__/CartNotification.test.tsx [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    );
+
+    fireEvent.click(screen.getByText("Close"));
+    expect(mockSetShow).toHaveBeenCalledWith(false);
+  });
+
+  it('opens cart in a new tab when "View my cart" button is clicked', () => {
+    // Mocking window.open
+    const mockOpen = vi.fn();
+    global.open = mockOpen;
```
Comment (bard) [inline packages/grant-explorer/src/features/common/__tests__/CartNotification.test.tsx:50]: Ouch. Not great but I get why it's needed. Can you set it as a mock in a `beforeEach` and make sure it's reset between tests?

File: packages/grant-explorer/src/features/common/__tests__/ErrorModal.test.tsx [hunk 4]
```diff
@@ -77,4 +77,32 @@ describe.skip("<ErrorModal />", () => {
     expect(setIsOpenFn).toBeCalledTimes(1);
     expect(setIsOpenFn).toBeCalledWith(false);
   });
+
+  it.skip("should close the modal if close on background click is enabled and background is clicked", () => {
```
Comment (bard) [inline packages/grant-explorer/src/features/common/__tests__/ErrorModal.test.tsx:81]: Should this stay skipped?

File: packages/grant-explorer/src/features/discovery/ApplyNowSection.tsx [hunk 5]
```diff
@@ -24,17 +24,23 @@ const ApplyNowSection = (props: ApplyNow) => {
...
+            {applyNowRoundsCount > 0 && `View All (${applyNowRoundsCount})`}
           </Link>
         </div>
       </div>
       {props.isLoading ? (
         <div className="grid grid-cols-1 sm:grid-cols-2 md:grid-cols-3 md:gap-6 2xl:grid-cols-4">
           {new Array(3).fill(undefined).map((_, index) => {
             return (
-              <Skeleton key={index} my={3} rounded={"md"} w={"full"} h={"64"} />
+              // @ts-expect-error Tsc too weak to represent this type lol
```
Comment (bard) [inline packages/grant-explorer/src/features/discovery/ApplyNowSection.tsx:35]: 🤣

File: packages/grant-explorer/src/features/round/__tests__/ViewRoundPage.test.tsx [hunk 6]
```diff
@@ -1,58 +1,75 @@
...
-  useSigner: () => mockSigner,
-  useNetwork: () => mockNetwork,
-  useSwitchNetwork: () => mockSwitchNetwork,
-}));
-
-jest.mock("react-router-dom", () => ({
-  ...jest.requireActual("react-router-dom"),
-  useParams: useParamsFn,
-}));
+fetchMock.mockIf(/summary/, JSON.stringify({}));
```
Comment (bard) [inline packages/grant-explorer/src/features/round/__tests__/ViewRoundPage.test.tsx:15]: I noticed `msw` in `package.json`, are we using only one of `fetch-mock` and `msw` or both?

**Review by bard — CHANGES_REQUESTED** [review]

File: packages/grant-explorer/src/cartStore.test.tsx [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+      .getState()
+      .projects.find(
+        (p) => p.grantApplicationId === project.grantApplicationId
+      );
+    expect(updatedProject?.amount).toBe(updatedAmount);
+  });
+
+  test("should set payout token for a specific chain", () => {
+    const chainId: ChainId = ChainId.MAINNET; // Mock ChainId
+    const payoutToken: PayoutToken = payoutTokensMap[ChainId.MAINNET][0];
```
[the remark this answers is not in the document]
Comment (vacekj) [reply packages/grant-explorer/src/cartStore.test.tsx:68]: True, ack'ed.

File: packages/grant-explorer/src/cartStore.test.tsx [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+    const nonExistingId = "1234"; // Mock ID
+    const initialProjects = [...useCartStorage.getState().projects];
+
+    useCartStorage.getState().remove(nonExistingId);
+
+    // Assert that the store state remains unchanged
+    expect(useCartStorage.getState().projects).toEqual(initialProjects);
+  });
+
+  test("should handle adding a project with duplicate ID gracefully", () => {
```
[the remark this answers is not in the document]
Comment (vacekj) [reply packages/grant-explorer/src/cartStore.test.tsx:147]: i can't rule out a bug that would enable someone to try to adding the same project twices, so I would rather guard against it.

File: packages/grant-explorer/src/features/api/voting.ts [hunk 9]
```diff
@@ -13,6 +22,25 @@ export type PermitSignature = {
...
 };
 
+export type PermitType = "dai" | "eip2612";
+/** Given a payout token, selects the correct permit type.
+ * - DAI is the old permit type without `value` and with the `allowed` prop
+ * - eip2612 is the standard permit interface, as specified in https://eips.ethereum.org/EIPS/eip-2612
+ *
+ * Old DAI permit type is only implemented on Ethereum and Polygon PoS. Check /docs/DAI.md for more info.
+ * */
+export const selectPermitType = (token: PayoutToken): PermitType => {
```
[the remark this answers is not in the document]
Comment (vacekj) [reply packages/grant-explorer/src/features/api/voting.ts:32]: Sure, will fix.

File: packages/grant-explorer/src/features/api/voting.ts (hunk 2 above)
Comment (vacekj) [reply packages/grant-explorer/src/features/api/voting.ts:257]: This cast is not actually even necessary here. Good catch.

File: packages/grant-explorer/src/features/api/voting.ts (hunk 1 above)
Comment (vacekj) [reply packages/grant-explorer/src/features/api/voting.ts:34]: We have several lists of payout tokens scattered accross the codebase, so a good ticket would be to merge them into one and encode this information there.

File: packages/grant-explorer/src/features/common/__tests__/CartNotification.test.tsx (hunk 3 above)
Comment (vacekj) [reply packages/grant-explorer/src/features/common/__tests__/CartNotification.test.tsx:50]: Yes, will fix.

File: packages/grant-explorer/src/features/common/__tests__/ErrorModal.test.tsx (hunk 4 above)
Comment (vacekj) [reply packages/grant-explorer/src/features/common/__tests__/ErrorModal.test.tsx:81]: I wasn't able to test this part of the modal. The funcitonality that it's testing is basically that we pass the "close on background click" prop correctly from our component to the headlessui component. I would welcome a better way to test just that, instead of testing the implementation detail of the actual headless ui component, which we have no business (lol) with.

File: packages/grant-explorer/src/features/discovery/ApplyNowSection.tsx (hunk 5 above)
Comment (vacekj) [reply packages/grant-explorer/src/features/discovery/ApplyNowSection.tsx:35]: wen `stc`

File: packages/grant-explorer/src/features/common/__tests__/ErrorModal.test.tsx (hunk 4 above)
Comment (bard) [reply packages/grant-explorer/src/features/common/__tests__/ErrorModal.test.tsx:81]: Ok, just double-checking because I've sometimes marked tests as skip/only during development then forgotten to un-mark them before pushing.

File: packages/grant-explorer/src/features/round/__tests__/ViewRoundPage.test.tsx (hunk 6 above)
Comment (vacekj) [reply packages/grant-explorer/src/features/round/__tests__/ViewRoundPage.test.tsx:15]: i deprecated msw in favour of fetch-mock as it was only used in one page and was giving some trouble. I'd prefer we do everything thru msw, as it more closely reflects the real-world scenario. 
So here, it's jsut a leftover in package.json. Removed.

File: docs/DAI.md [hunk 10]
```diff
@@ -0,0 +1,24 @@
+# DAI in relation to Allo and Grants Stack
```
[the remark this answers is not in the document]
Comment (bard) [reply docs/DAI.md:1]: Are there parts of the source that are explained by the content of this document? If so, we could add something like `// to find more about why do this, see docs/DAI.md`. Could save chronic non-readers like yours truly.

File: packages/grant-explorer/src/features/api/voting.ts (hunk 1 above)
Comment (vacekj) [reply packages/grant-explorer/src/features/api/voting.ts:34]: https://github.com/gitcoinco/grants-stack/issues/2271

File: docs/DAI.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (vacekj) [reply docs/DAI.md:1]: ah, I see your point now - it's already in the docs for the getPermitType fn.
image

File: packages/grant-explorer/src/features/round/__tests__/PassportConnect.test.tsx [hunk 11]
```diff
@@ -146,8 +147,8 @@ describe("<PassportConnect/>", () => {
         json: () => mockJsonPromise,
       } as unknown as Response;
 
-      (fetchPassport as jest.Mock).mockResolvedValueOnce(mockPassportPromise);
-      (submitPassport as jest.Mock).mockResolvedValueOnce(jest.fn());
+      (fetchPassport as Mock).mockResolvedValueOnce(mockPassportPromise);
+      (submitPassport as Mock).mockResolvedValueOnce(vi.fn());
```
Comment (bard) [inline packages/grant-explorer/src/features/round/__tests__/PassportConnect.test.tsx:151]: Not part of this PR, but this looks slightly suspicious — does `submitPassport` actually return a function?

File: packages/grant-explorer/src/store.ts [hunk 12]
```diff
@@ -71,6 +79,10 @@ export const useCartStorage = create<CartState>()(
       },
       chainToPayoutToken: ethOnlyPayoutTokens,
       setPayoutTokenForChain: (chainId: ChainId, payoutToken: PayoutToken) => {
+        if (!Object.values(ChainId).includes(chainId)) {
```
Comment (bard) [inline packages/grant-explorer/src/store.ts:82]: Is setting an invalid payout token for a chain a valid operation? If not, we might want some degree of warning on the `console.warn` to `throw` spectrum.

**Review by bard — APPROVED** [review]

File: packages/grant-explorer/src/features/contributors/__tests__/ViewContributionHistory.test.tsx [hunk 13]
```diff
@@ -89,25 +91,38 @@ const breadCrumbs = [
...
-  useEnsName: jest.fn().mockReturnValue({ data: "" }),
-  useAccount: jest.fn().mockReturnValue({ data: "mockedAccount" }),
+vi.mock("../../common/Navbar");
+vi.mock("../../common/Auth");
+
+vi.mock("@rainbow-me/rainbowkit", () => ({
+  ConnectButton: vi.fn(),
 }));
 
+vi.mock("react-router-dom", async () => {
```
[the remark this answers is not in the document]
Comment (vacekj) [reply packages/grant-explorer/src/features/contributors/__tests__/ViewContributionHistory.test.tsx:100]: True, but I think that doesn't invalidate the need for this mock?

File: packages/grant-explorer/src/features/api/voting.ts [hunk 14]
```diff
@@ -65,7 +84,7 @@ export const voteUsingMRCContract = async (
   } else if (permit) {
     /* Is token DAI? */
     /** DAI on optimism supports normal eip2612 permit, so we skip that*/
```
[the remark this answers is not in the document]
Comment (vacekj) [reply packages/grant-explorer/src/features/api/voting.ts]: This comment no longer exists in the code.

File: packages/grant-explorer/src/checkoutStore.test.tsx [hunk 15]
```diff
@@ -0,0 +1,70 @@
+import { useCheckoutStore } from "./checkoutStore";
+import { ProgressStatus } from "./features/api/types";
+import { ChainId } from "common";
+import { beforeEach } from "vitest";
+
+const store = useCheckoutStore;
```
[the remark this answers is not in the document]
Comment (vacekj) [reply packages/grant-explorer/src/checkoutStore.test.tsx:6]: it's a hook if you call it, but there's other method on the bare object that you can use. The convention is to call it useXStore for this reason.

File: packages/grant-explorer/src/context/__tests__/RoundContext.test.tsx [hunk 16]
```diff
@@ -5,15 +5,12 @@ import { makeRoundData } from "../../test-utils";
 import { getRoundById } from "../../features/api/round";
 import { Round } from "../../features/api/types";
 
-jest.mock("../../features/api/round");
-jest.mock("wagmi");
-jest.mock("@rainbow-me/rainbowkit", () => ({
-  ConnectButton: jest.fn(),
-}));
+vi.mock("../../features/api/round");
+vi.mock("wagmi");
```
[the remark this answers is not in the document]
Comment (vacekj) [reply packages/grant-explorer/src/context/__tests__/RoundContext.test.tsx:9]: added as a todo in code.

File: packages/grant-explorer/src/features/common/__tests__/CartNotification.test.tsx [hunk 17]
```diff
@@ -0,0 +1,93 @@
+import React from "react";
+import { render, fireEvent, screen } from "@testing-library/react";
+import CartNotification, { ProjectView } from "../CartNotification";
+import { makeApprovedProjectData } from "../../../test-utils";
+import { beforeEach, vi } from "vitest"; // Adjust the path accordingly
+import DefaultLogoImage from "../../../assets/default_logo.png";
+
+const mockProject = makeApprovedProjectData();
```
[the remark this answers is not in the document]
Comment (vacekj) [reply packages/grant-explorer/src/features/common/__tests__/CartNotification.test.tsx:8]: side-effects bad 🫡, fixed.

File: packages/grant-explorer/src/features/common/__tests__/CopyToClipboardButton.test.tsx [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+
+    expect(screen.getByText("Link Copied")).toBeInTheDocument();
+  });
+
+  it("should copy value to clipboard when clicked", () => {
+    renderWithContext(<CopyToClipboardButton textToCopy={textToCopy} />);
+    const copyButton = screen.getByRole("button");
+    fireEvent.click(copyButton);
+
+    expect(navigator.clipboard.readText()).toEqual(textToCopy);
```
[the remark this answers is not in the document]
Comment (vacekj) [reply packages/grant-explorer/src/features/common/__tests__/CopyToClipboardButton.test.tsx:41]: Unless this is a rhetorical question, I believe that it is?

File: packages/grant-explorer/src/features/round/__tests__/PassportConnect.test.tsx (hunk 11 above)
Comment (vacekj) [reply packages/grant-explorer/src/features/round/__tests__/PassportConnect.test.tsx:151]: Not sure, but OOS for this PR.

File: packages/grant-explorer/src/store.ts (hunk 12 above)
Comment (vacekj) [reply packages/grant-explorer/src/store.ts:82]: It's an invalid op. There is no use-case where this should happen. Added a console.warn.

File: packages/grant-explorer/src/features/contributors/__tests__/ViewContributionHistory.test.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (bard) [reply packages/grant-explorer/src/features/contributors/__tests__/ViewContributionHistory.test.tsx:100]: There would be no need to mock as `useNavigate` and `useParams` would be retrieved from context. The in-memory URL would have to be set as part of the rendering.

File: packages/grant-explorer/src/features/common/__tests__/CopyToClipboardButton.test.tsx (hunk 18 above)
[the remark this answers is not in the document]
Comment (bard) [reply packages/grant-explorer/src/features/common/__tests__/CopyToClipboardButton.test.tsx:41]: I mean, `readText` is defined in the test itself and then asserted in the test itself, doesn't exercise the component at all unless I'm missing something. Is there any modification you could do to the `<CopyToClipboardButton>` that would cause that test to fail?

Something like this might be closer to the spirit of the test:

```tsx
    const clipboard = { writeText: vi.fn() }
    Object.assign(navigator, { clipboard })
    renderWithContext(<CopyToClipboardButton textToCopy={'foobar'} />);
    const copyButton = screen.getByRole("button");
    fireEvent.click(copyButton);
    expect(navigator.clipboard.writeText()).toHaveBeenCalledWith('foobar');
```

**Review by bard — APPROVED** [review]

**Review by michellema1208 — APPROVED** [review]
